app/electron/main/code/components.ts

```diff
@@ -0,0 +1,139 @@
+import {
+    Project,
+    ts,
+    Node,
+    FunctionDeclaration,
+    ClassDeclaration,
+    VariableStatement,
+} from 'ts-morph';
+
+import * as path from 'path';
+import { promises as fs } from 'fs';
+
+function isUppercase(s: string) {
+    return s === s.toUpperCase();
+}
+
+export interface ReactComponentDescriptor {
+    name: string;
+    sourceFilePath: string;
+}
+
+function isExported(node: FunctionDeclaration | VariableStatement | ClassDeclaration) {
+    return node.getModifiers().some((m) => m.getKind() === ts.SyntaxKind.ExportKeyword);
+}
+
+function getReactComponentDescriptor(
+    node: Node,
+): Omit<ReactComponentDescriptor, 'sourceFilePath'> | null {
+    if (Node.isVariableStatement(node)) {
+        if (!isExported(node)) {
+            return null;
+        }
+
+        const declaration = node.getDeclarationList().getDeclarations().at(0);
+        if (declaration == null) {
+            return null;
+        }
+
+        const name = declaration.getName();
+        if (name == null) {
+            return null;
+        }
+
+        const initializer = declaration.getInitializer();
+        if (Node.isArrowFunction(initializer)) {
+            return { name };
+        }
+    }
+
+    if (Node.isFunctionDeclaration(node)) {
+        if (!isExported(node)) {
+            return null;
+        }
+        const name = node.getName();
+
+        if (name != null && isUppercase(name[0])) {
+            return { name };
+        }
+    }
+
+    if (Node.isClassDeclaration(node)) {
+        if (!isExported(node)) {
+            return null;
+        }
+        const heritageClauses = node.getHeritageClauses();
+        if (heritageClauses == null) {
+            return null;
+        }
+        const isDerivedFromReactComponent = heritageClauses.some(
+            (clause) =>
+                clause.getText().includes('React.Component') ||
+                clause.getText().includes('React.PureComponent'),
+        );
+
+        if (!isDerivedFromReactComponent) {
+            return null;
+        }
+
+        const name = node.getName();
+        if (name == null) {
+            return null;
+        }
+
+        return { name };
+    }
+
+    return null;
+}
+
+function extractReactComponentsFromFile(filePath: string) {
+    const project = new Project();
+    const sourceFile = project.addSourceFileAtPath(filePath);
+
+    const exportedComponents: ReactComponentDescriptor[] = [];
+
+    sourceFile.forEachChild((node) => {
+        const descriptor = getReactComponentDescriptor(node);
+        if (descriptor != null) {
+            exportedComponents.push({
+                ...descriptor,
+                sourceFilePath: sourceFile.getFilePath(),
+            });
+        }
+    });
+
+    return exportedComponents;
+}
+
+async function scanDirectory(dir: string): Promise<string[]> {
```

Comment:
Question: Should we be ignoring node_modules? Seemingly no because libs are all there. 

But maybe we should be explicitly choosing certain libs because this seems to pick up every single React class in node_modules. Maybe it's just organizing them.

Reply:
great idea, implemented in https://github.com/onlook-dev/onlook/pull/285/commits/1b73457761da4d3270da7c11235cf5d4df0adcbb